### PR TITLE
Feature/1577 caddisfly question

### PR DESCRIPTION
**Pull request by mtwestra**
Pull request for caddisfly integration.

### CODE REVIEW

**Comment by mtwestra** [comment]
Issues: #1643, #1577, #1570

File: GAE/src/com/gallatinsystems/survey/domain/xml/Question.java [hunk 1]
```diff
@@ -348,6 +354,24 @@ public String getCascadeResource() {
     public void setCascadeResource(String value) {
         this.cascadeResource = value;
     }
+    
+    /**
+     * Gets the value of the cascadeResource property.
```
Comment (stellanl) [inline GAE/src/com/gallatinsystems/survey/domain/xml/Question.java]: Copypasted comment not updated

File: GAE/src/org/waterforpeople/mapping/app/web/rest/CaddisflyResourceRestService.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/*  Copyright (C) 2014-2015 Stichting Akvo (Akvo Foundation)
```
Comment (stellanl) [inline GAE/src/org/waterforpeople/mapping/app/web/rest/CaddisflyResourceRestService.java]: Copyright year not updated

File: GAE/src/org/waterforpeople/mapping/domain/CaddisflyResource.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
+/*  Copyright (C) 2014 Stichting Akvo (Akvo Foundation)
```
Comment (stellanl) [inline GAE/src/org/waterforpeople/mapping/domain/CaddisflyResource.java]: Copyright year not updated

File: GAE/src/org/waterforpeople/mapping/domain/CaddisflyResult.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
+/*  Copyright (C) 2014 Stichting Akvo (Akvo Foundation)
```
Comment (stellanl) [inline GAE/src/org/waterforpeople/mapping/domain/CaddisflyResult.java]: Copyright year not updated

**Comment by stellanl** [comment]
I found three non-updated copyright years, and another comment with old copypasted info, but nothing else.

**Comment by mtwestra** [comment]
### Test plan

1) go to flowaglimmerofhope dashboard
2) create a new survey and a new question
3) choose the question type 'caddisfly'
4) publish the survey and assign it to a Flow phone
5) download and install the caddisfly app
6) open the survey on the Flow phone and start a new data point
7) The caddisfly question should be displayed correctly and when pressed, open the caddisfly app.


###Second test
1) go to flowaglimmerofhope dashboard
2) assign the survey 1.Mark/Caddisfly test to a phone (surveyid = 37564000)
3) open the flow app, start a new survey
4) fill all of the questions using the strip test solution or sensor. (if not possible, just do the ones you can)
5) upload the data
6) go to the data tab, and check the data is displayed correctly
7) go to the monitoring tab, and check the data is displayed correctly
8) export the data using 'export for external systems', and check the data is exported correctly.

**Comment by janagombitova** [comment]
@mtwestra and @muloem I ran the first set of UAT tests for these issues. Here are my comments and findings (I will try to keep the focus on Flow related points, we can cover issues around the Caddisfly app in their repository): 

**When creating a survey with multiple Caddisfly questions** 

For the 1st Caddisfly question the tests dropdown opens on 'Select test'. However, the following Caddisfly question opens the test dropdown on the previously selected test and not on 'Select test'. 

caddisfly dropdown2

**In the app**

I could not run any test in the beginning as after clicking on "prepare for test" the Caddisfly app asked me to install OpenCV manager. Is this a compulsory requirement? I think the tests should run without the user having the install another (3rd) app on her phone as it does not make water quality testing easier. From my understanding this is on the Caddisfly side of things and should be part of the next improvement steps in my opinion. 

**After installing OpenCV manager** 

If I had a Caddisfly question in a repeated question group, I was not able to run the test as the app said it Cannot start test. This happened for both EC and chlorine which worked in other question groups. 

screenshot_2016-08-03-15-36-27
 
After a test is successfully completed the app opens Flow nicely, but in the answer field we only show: "Test captured successfully" In my point of view, it would be useful to show the value captured. It does not have to be the full image and value as shown in the Data tab, but the value itself will tell the user more about the water sample. 

**Back in the dashboard after data is submitted**

In the Maps (public and private) in the side bar the answer is shown as a json string. 

screen shot 2016-08-03 at 15 23 08

**Reports**
I could not check how the reports work.

**Comment by mtwestra** [comment]
1. When creating a survey with multiple Caddisfly questions
This should be easy to fix.

2. In the app
Yes, unfortunately OpenCV is a hard requirement. Both the strip test code and the fluoride code use it. It only needs to be installed once, of course.

3. After installing OpenCV manager
No idea why that is (the repeated question group not working). Need to investigate. Could you send me the survey xml file perhaps?

4. on showing the result in FLOW app
Yes, good idea. We wanted to do that, but Inigo didn't have the time anymore as it is quite complex. It will need to wait until we have a FLOW app developer again

5. Back in the dashboard after data is submitted
Yes, the data is not handled in the map yet. It would not be very hard to do though - we can reuse the code for the display in the inspect data tab

**Comment by janagombitova** [comment]
> When creating a survey with multiple Caddisfly questions
> This should be easy to fix.

Ok, would be good if you can tackle it for this round 

> In the app
> Yes, unfortunately OpenCV is a hard requirement. Both the strip test code and the fluoride code use it. It only needs to be installed once, of course.

I see, then we need to add this to the support manual and perhaps also to the Google Play Store copy, so users download everything before they head off to the field.

> After installing OpenCV manager
> No idea why that is (the repeated question group not working). Need to investigate. Could you send me the survey xml file perhaps?

Would be also great if this can be fixed for this round. I used flowaglimmerofhope > Jana > 1.9.10 test Caddisfly > form ID: 50234000

> on showing the result in FLOW app
> Yes, good idea. We wanted to do that, but Inigo didn't have the time anymore as it is quite complex. It will need to wait until we have a FLOW app developer again

I will create a separate issue for this in Flow-mobile repository 

> Back in the dashboard after data is submitted
> Yes, the data is not handled in the map yet. It would not be very hard to do though - we can reuse the code for the display in the inspect data tab

I think we need to create a separate issue for this one as well and we can plan it in for a later milestone as we will need to cover the public and private maps, Mapbox maps and CartoDB maps. Will create the issue

**Comment by mtwestra** [comment]
The only remaining points here are getting rid of openCV, and displaying the result in the FLOW app in a more meaningful way.

File: Dashboard/app/js/lib/views/views-public.js [hunk 5]
```diff
@@ -52,13 +52,85 @@ Ember.Handlebars.registerHelper('tooltip', function (i18nKey) {
   );
 });
 
+function renderCaddisflyAnswer(json){
+  name = ""
+  imageUrl = ""
+  result = Ember.A();
```
Comment (iperdomo) [inline Dashboard/app/js/lib/views/views-public.js]: `name`, `imageUrl` and `result` without a `var` statement, creates **global** variables

File: Dashboard/app/js/lib/views/views-public.js [hunk 6]
```diff
@@ -52,13 +52,85 @@ Ember.Handlebars.registerHelper('tooltip', function (i18nKey) {
   );
 });
 
+function renderCaddisflyAnswer(json){
```
Comment (iperdomo) [inline Dashboard/app/js/lib/views/views-public.js]: `renderCaddisflyAnswer` should me a member of the `FLOW` object

File: Dashboard/app/js/lib/views/views-public.js [hunk 7]
```diff
@@ -52,13 +52,85 @@ Ember.Handlebars.registerHelper('tooltip', function (i18nKey) {
...
+      });
+    }
+    // get out image url
+    if (!Ember.empty(jsonParsed.image)){
+      imageUrl = FLOW.Env.photo_url_root + jsonParsed.image.trim();
+    }
+
+    // contruct html
+    html = "<div><strong>" + name + "</strong></div>"
+    result.forEach(function(item){
```
Comment (iperdomo) [inline Dashboard/app/js/lib/views/views-public.js]: It seems that we're inconsistently use `forEach` or `map` + `join`. It feels more _clean_ to use map + join then modifying a closed variable `html` in the outer scope.

File: Dashboard/app/js/lib/views/views-public.js [hunk 8]
```diff
@@ -67,6 +139,39 @@ Ember.Handlebars.registerHelper('placemarkDetail', function () {
...
+    YYYY-mm-dd */
+function renderTimeStamp(timestamp) {
+  var d, t, date, month, year;
+  t = parseInt(timestamp, 10);
+  if (isNaN(t)) {
+    return "";
+  }
+
+  d = new Date(t);
+  if (d) {
```
Comment (iperdomo) [inline Dashboard/app/js/lib/views/views-public.js]: Instead of having a big `if` with an small `else` you can just `return ""` if `!d`

File: Dashboard/app/js/lib/views/views.js [hunk 9]
```diff
@@ -87,6 +87,46 @@ Ember.Handlebars.registerHelper('tooltip', function (i18nKey) {
 });
 
 
+function renderCaddisflyAnswer(json){
```
Comment (iperdomo) [inline Dashboard/app/js/lib/views/views.js]: This function should be part of the `FLOW` namespace/object

File: Dashboard/app/js/lib/views/views.js [hunk 10]
```diff
@@ -87,6 +87,46 @@ Ember.Handlebars.registerHelper('tooltip', function (i18nKey) {
...
+  name = ""
+  imageUrl = ""
+  result = Ember.A();
+  if (!Ember.empty(json)){
```
Comment (iperdomo) [inline Dashboard/app/js/lib/views/views.js]: `name`, `imageUrl`, `result` are **globals** without a `var` statement

File: Dashboard/app/js/lib/views/views.js (hunk 9 above)
Comment (iperdomo) [inline Dashboard/app/js/lib/views/views.js]: Why do we need a copy of the function in a different file?

File: Dashboard/app/js/templates/navData/monitoring-data.handlebars [hunk 11]
```diff
@@ -157,9 +157,15 @@
                             {{else}}
                                 {{t _no_signature_found}}
                             {{/if}}
-                        {{else}}
+                        {{else}} {{#if view.isCaddisflyType}}
```
Comment (iperdomo) [inline Dashboard/app/js/templates/navData/monitoring-data.handlebars]: Indentation of this section is not ideal. There is no easy way to _scan_ the code for the if` block

File: Dashboard/app/js/templates/navData/question-answer.handlebars [hunk 12]
```diff
@@ -55,9 +55,15 @@
                     {{else}}
                         {{t _no_signature_found}}
                     {{/if}}
-                {{else}}
+                {{else}} {{#if view.isCaddisflyType}}
```
Comment (iperdomo) [inline Dashboard/app/js/templates/navData/question-answer.handlebars]: Idem

File: GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
+/*
+ *  Copyright (C) 2014 Stichting Akvo (Akvo Foundation)
```
Comment (iperdomo) [inline GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java]: Wrong year

Comment (iperdomo) [inline GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java]: This file is not properly formatted. We're using Eclipse's default format

File: GAE/src/org/waterforpeople/mapping/domain/CaddisflyResult.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+		private String unit;
+		private int id;
+
+		public CaddisflyResult(String name, String unit, int id){
+			this.setName(name);
+			this.setUnit(unit);
+			this.setId(id);
+		}
+
+		public CaddisflyResult(){
```
Comment (iperdomo) [inline GAE/src/org/waterforpeople/mapping/domain/CaddisflyResult.java]: Delete if not used?

File: GAE/src/resources/caddisfly/caddisfly-tests.json [hunk 15]
```diff
@@ -0,0 +1,154 @@
+{
```
Comment (iperdomo) [inline GAE/src/resources/caddisfly/caddisfly-tests.json]: AFAIK, this is a shared resource between the app and the server. Shouldn't we include some time of `"version"` property so we know we're using the same file? I'm not completely sure what are the implications of having out-of-sync files (e.g. the app behind the server)

File: Dashboard/app/js/lib/views/data/question-answer-view.js [hunk 16]
```diff
@@ -157,6 +161,81 @@ FLOW.QuestionAnswerView = Ember.View.extend({
...
+  			  newResult = {
+  			    "name":item.name,
+  				  "value":item.value,
+  				  "unit":item.unit,
+  	  		};
+  		  	result.push(newResult);
+  		  });
+  	  }
+    }
+    this.set('testResult',result);
```
Comment (iperdomo) [inline Dashboard/app/js/lib/views/data/question-answer-view.js]: Could this whole method be reduced to just parsing the string and passing it to `Ember.A()` ? e.g. `Ember.A([{"name":"name1", "value": "some value", "unit": "unit"}])`

File: Dashboard/app/js/lib/views/views.js (hunk 9 above)
Comment (mtwestra) [inline Dashboard/app/js/lib/views/views.js]: good question. The fact that there are two files is that one is loaded when the user is logged in, and the other is loaded when the user is not logged in. However, the common elements should be able to be separated out and put in a single file.

File: Dashboard/app/js/lib/views/views-public.js (hunk 6 above)
Comment (mtwestra) [inline Dashboard/app/js/lib/views/views-public.js]: done

File: Dashboard/app/js/lib/views/views-public.js (hunk 5 above)
Comment (mtwestra) [inline Dashboard/app/js/lib/views/views-public.js]: done

File: Dashboard/app/js/lib/views/views-public.js (hunk 7 above)
Comment (mtwestra) [inline Dashboard/app/js/lib/views/views-public.js]: done

File: Dashboard/app/js/lib/views/views-public.js (hunk 8 above)
Comment (mtwestra) [inline Dashboard/app/js/lib/views/views-public.js]: done

File: Dashboard/app/js/lib/views/views.js (hunk 10 above)
Comment (mtwestra) [inline Dashboard/app/js/lib/views/views.js]: done

File: Dashboard/app/js/templates/navData/monitoring-data.handlebars (hunk 11 above)
Comment (mtwestra) [inline Dashboard/app/js/templates/navData/monitoring-data.handlebars]: I think it is not that bad actually - all the 'else' blocks are aligned. This should really be a 'case' block, but that is not available in handlebar files. The alternative is an 8-fold indented block, which I am not sure would improve it.

File: Dashboard/app/js/templates/navData/question-answer.handlebars (hunk 12 above)
Comment (mtwestra) [inline Dashboard/app/js/templates/navData/question-answer.handlebars]: idem, see above

File: GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java (hunk 13 above)
Comment (mtwestra) [inline GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java]: year: fixed
I've turned on formatting now - it was not being done automatically.

File: GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+     */
+    public List<CaddisflyResource> listResources() {
+        List<CaddisflyResource> result = new ArrayList<CaddisflyResource>();
+    	
+        try {
+        	InputStream stream = getClass().getClassLoader().getResourceAsStream ("resources/caddisfly/caddisfly-tests.json");
+        	String jsonTxt = IOUtils.toString(stream);
+        	
+            // create a list of caddisflyResource objects
+        	ObjectMapper mapper = new ObjectMapper();
```
Comment (mtwestra) [inline GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java]: done

File: GAE/src/com/gallatinsystems/survey/domain/Question.java [hunk 18]
```diff
@@ -451,4 +452,12 @@ public Long getSourceQuestionId() {
     public void setSourceQuestionId(Long originalQuestionId) {
         this.sourceQuestionId = originalQuestionId;
     }
+
+	public String getCaddisflyResourceUuid() {
```
Comment (mtwestra) [inline GAE/src/com/gallatinsystems/survey/domain/Question.java]: done

File: GAE/src/org/waterforpeople/mapping/app/web/rest/CaddisflyResourceRestService.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Inject
+    private CaddisflyResourceDao caddisflyResourceDao;
+
+    @RequestMapping(method = RequestMethod.GET, value = "")
+    @ResponseBody
+    public Map<String, List<CaddisflyResource>> listCaddisflyResources() {
+        final Map<String, List<CaddisflyResource>> response = new HashMap<String, List<CaddisflyResource>>();
+        List<CaddisflyResource> caddisList = caddisflyResourceDao.listResources();
+        // TODO create a keyId for each, because it is needed in the dashboard.
```
Comment (mtwestra) [inline GAE/src/org/waterforpeople/mapping/app/web/rest/CaddisflyResourceRestService.java]: removed - already implemented

File: GAE/src/org/waterforpeople/mapping/dataexport/GraphicalSurveySummaryExporter.java [hunk 20]
```diff
@@ -693,11 +719,16 @@ private void writeAnswer(
...
 
             case OPTION:
-                Long qId = questionDto.getKeyId();
+                qId = questionDto.getKeyId();
                 cells.addAll(optionCellValues(questionDto.getKeyId(), value, useQuestionId,
                         optionMap.get(qId), allowOtherMap.get(qId)));
                 break;
 
+            case CADDISFLY:
+            	qId = questionDto.getKeyId();
```
Comment (mtwestra) [inline GAE/src/org/waterforpeople/mapping/dataexport/GraphicalSurveySummaryExporter.java]: done

File: GAE/src/org/waterforpeople/mapping/dataexport/GraphicalSurveySummaryExporter.java [hunk 21]
```diff
@@ -761,6 +804,127 @@ private static String dateCellValue(String value) {
...
     }
 
+    /*
+     * Validates the structure of a caddisfly value
+     * Returns true if the structure is correct, false if not.
+     */
+    @SuppressWarnings("unchecked")
+	private static Boolean validateCaddisflyValue(String value,
+			Integer numResults, Boolean hasImage) {
+    	try {
```
Comment (mtwestra) [inline GAE/src/org/waterforpeople/mapping/dataexport/GraphicalSurveySummaryExporter.java]: done

File: GAE/src/org/waterforpeople/mapping/dataexport/GraphicalSurveySummaryExporter.java [hunk 22]
```diff
@@ -761,6 +804,127 @@ private static String dateCellValue(String value) {
...
+		}
+
+		return true;
+	}
+
+    /*
+     * Creates the cell values for a caddisfly question.
+     */
+    @SuppressWarnings("unchecked")
+	private List<String> caddisflyCellValues(Long questionId, String value, Integer numResults, Boolean hasImage, String imagePrefix){
```
Comment (mtwestra) [inline GAE/src/org/waterforpeople/mapping/dataexport/GraphicalSurveySummaryExporter.java]: done

File: GAE/src/org/waterforpeople/mapping/dataexport/GraphicalSurveySummaryExporter.java [hunk 23]
```diff
@@ -997,7 +1162,8 @@ private String sanitize(String s) {
      *         representing all the non-summarizable question Ids (i.e. those that aren't OPTION or
      *         NUMBER questions)
      */
-    protected Object[] createRawDataHeader(Workbook wb, Sheet sheet,
+    @SuppressWarnings("unchecked")
+	protected Object[] createRawDataHeader(Workbook wb, Sheet sheet,
```
Comment (mtwestra) [inline GAE/src/org/waterforpeople/mapping/dataexport/GraphicalSurveySummaryExporter.java]: sorry, yes. Indentation is hard, apparently... :-(

File: GAE/src/org/waterforpeople/mapping/domain/CaddisflyResource.java [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+		this.setUuid(uuid);
+		this.setSubtype(subtype);
+		this.setBrand(brand);
+		this.setTags(tags);
+		this.setDescription(description);
+		this.setNumResults(numResults);
+		this.setHasImage(hasImage);
+	}
+
+	public CaddisflyResource(){
```
Comment (mtwestra) [inline GAE/src/org/waterforpeople/mapping/domain/CaddisflyResource.java]: removed. Apparently I started it twice.

File: GAE/src/org/waterforpeople/mapping/domain/CaddisflyResult.java [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ *  The full license text can also be seen at <http://www.gnu.org/licenses/agpl.html>.
+ */
+
+package org.waterforpeople.mapping.domain;
+
+import java.io.Serializable;
+import java.util.List;
+
+public class CaddisflyResult implements Serializable,Comparable {
```
Comment (mtwestra) [inline GAE/src/org/waterforpeople/mapping/domain/CaddisflyResult.java]: done

File: GAE/src/org/waterforpeople/mapping/domain/CaddisflyResult.java (hunk 14 above)
Comment (mtwestra) [inline GAE/src/org/waterforpeople/mapping/domain/CaddisflyResult.java]: idem - deleted.

File: GAE/src/resources/caddisfly/caddisfly-tests.json (hunk 15 above)
Comment (mtwestra) [inline GAE/src/resources/caddisfly/caddisfly-tests.json]: I don't think it is necessary - the app is designed to fail gracefully - it will just report 'unknown test'. If we had a version number it would basically do the same thing.

File: Dashboard/app/js/lib/views/data/question-answer-view.js (hunk 16 above)
Comment (mtwestra) [inline Dashboard/app/js/lib/views/data/question-answer-view.js]: I don't think so, because there can be multiple results, so the array will be filled with objects, each having a name, value and unit.

Comment (mtwestra) [inline Dashboard/app/js/lib/views/data/question-answer-view.js]: done

File: Dashboard/app/js/lib/views/views.js [hunk 26]
```diff
@@ -87,6 +87,47 @@ Ember.Handlebars.registerHelper('tooltip', function (i18nKey) {
...
+      });
+    }
+    // get out image url
+    if (!Ember.empty(jsonParsed.image)){
+      imageUrl = FLOW.Env.photo_url_root + jsonParsed.image.trim();
+    }
+
+    // contruct html
+    html = "<div><strong>" + name + "</strong></div>"
+    html += result.map(function(item){
```
Comment (iperdomo) [inline Dashboard/app/js/lib/views/views.js]: It seems that there is no need to have an intermediate copy of the `jsonParsed.result` in `result` if that is already an array and can be used with `map`?

File: Dashboard/app/js/lib/views/views.js [hunk 27]
```diff
@@ -135,34 +178,33 @@ Ember.Handlebars.registerHelper('placemarkDetail', function () {
...
-      dateString = "0" + date.toString();
-    } else {
-      dateString = date.toString();
-    }
-
-    return year + "-" + monthString + "-" + dateString;
-  } else {
-    return "";
-  }
+	var d, t, date, month, year;
```
Comment (iperdomo) [inline Dashboard/app/js/lib/views/views.js]: Unrelated change, broke indentation

File: GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+	private static ObjectMapper mapper = new ObjectMapper();
+
+	private static final Logger log = Logger.getLogger(CascadeResourceDao.class
+			.getName());
+
+	/**
+	 * lists caddisfly resources. Source is the json file caddisfly-tests.json
+	 * stored in WEB-INF/resources
+	 * 
+	 * @param item
```
Comment (iperdomo) [inline GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java]: Remove the non documented parameter?

File: GAE/src/org/waterforpeople/mapping/app/gwt/client/survey/QuestionDto.java [hunk 29]
```diff
@@ -26,458 +26,466 @@
...
-    }
-
-    public void setAllowPolygon(Boolean allowPolygon) {
-        this.allowPolygon = allowPolygon;
-    }
-
-    public Boolean getAllowPolygon() {
-        return allowPolygon;
-    }
+	private static final long serialVersionUID = -4708385830894435407L;
```
Comment (iperdomo) [inline GAE/src/org/waterforpeople/mapping/app/gwt/client/survey/QuestionDto.java]: Unrelated changes broke indentation. You can always review your diff before committing and pushing

File: Dashboard/app/js/lib/views/data/question-answer-view.js [hunk 30]
```diff
@@ -53,7 +53,11 @@ FLOW.QuestionAnswerView = Ember.View.extend({
     return this.get('questionType') === 'SIGNATURE' || (this.content && this.content.get('type') === 'SIGNATURE');
   }.property('this.questionType'),
 
-  nonEditableQuestionTypes: ['GEO', 'PHOTO', 'VIDEO', 'GEOSHAPE', 'SIGNATURE'],
+  isCaddisflyType: function(){
```
Comment (mtwestra) [inline Dashboard/app/js/lib/views/data/question-answer-view.js]: I don't see the problem...

File: Dashboard/app/js/lib/views/data/question-answer-view.js [hunk 31]
```diff
@@ -157,6 +161,74 @@ FLOW.QuestionAnswerView = Ember.View.extend({
...
+  * Extracts the 'result' key from a json string, and parses the array of results.
+  * Creates an array with name, value, unit for each result
+  * Example JSON format: {"result":[{"name":"Total Chlorine (ppm)","value":10,
+  * "unit":"ppm","id":0},{"name":"Free Chlorine (ppm)","value":0.5,"unit":"ppm",
+  * "id":1}],"type":"caddisfly","name":"Chlorine and Free Chlorine",
+  * "uuid":"bf1c19c0-9788-4e26-999e-1b5c6ca28111","image":"b3893f16-6a02-4e92-a13e-fce25223a0c5.png"}
+  */
+  parseTestJson: function(){
+	  var c = this.content, testJson, newResult, image;
+  	result=Ember.A();
```
Comment (mtwestra) [inline Dashboard/app/js/lib/views/data/question-answer-view.js]: done

File: Dashboard/app/js/lib/views/data/question-answer-view.js [hunk 32]
```diff
@@ -157,6 +161,74 @@ FLOW.QuestionAnswerView = Ember.View.extend({
...
+   * "uuid":"bf1c19c0-9788-4e26-999e-1b5c6ca28111","image":"b3893f16-6a02-4e92-a13e-fce25223a0c5.png"}
+   * 
+   * Extracts the 'name' attribute from a Caddisfly JSON result string 
+   */
+  testName: function(){
+  	var c = this.content, testJson;
+  	if (c && c.get('value')) {
+  	  testJson = JSON.parse(c.get('value'));
+  	  if (!Ember.empty(testJson.result))
+  	  {
```
Comment (mtwestra) [inline Dashboard/app/js/lib/views/data/question-answer-view.js]: done

File: Dashboard/app/js/lib/views/surveys/question-view.js [hunk 33]
```diff
@@ -137,6 +138,11 @@ FLOW.QuestionView = FLOW.View.extend({
             || (this.type && this.type.get('value') === 'SIGNATURE');
   }.property('this.type'),
 
+  amCaddisflyType: function () {
```
Comment (mtwestra) [inline Dashboard/app/js/lib/views/surveys/question-view.js]: done

File: Dashboard/app/js/lib/views/surveys/question-view.js [hunk 34]
```diff
@@ -197,6 +206,16 @@ FLOW.QuestionView = FLOW.View.extend({
     	FLOW.selectedControl.set('selectedCascadeResource', cascadeResource);
     }
 
+    // reset selected caddisfly resource
+    FLOW.selectedControl.set('selectedCaddisflyResource', null);
+    // if the caddisflyResourceUuid is not null, get the resource
+    if (!Ember.empty(FLOW.selectedControl.selectedQuestion.get('caddisflyResourceUuid'))) {
+      FLOW.caddisflyResourceControl.get('content').forEach(function (item) {
+        if (item.get('uuid') == FLOW.selectedControl.selectedQuestion.get('caddisflyResourceUuid')) {
+          FLOW.selectedControl.set('selectedCaddisflyResource', item);
```
Comment (mtwestra) [inline Dashboard/app/js/lib/views/surveys/question-view.js]: Possibly, but I don't have time to find out. There are only about 10 of them or so.

File: Dashboard/app/js/lib/views/views.js [hunk 35]
```diff
@@ -87,6 +87,47 @@ Ember.Handlebars.registerHelper('tooltip', function (i18nKey) {
...
+    jsonParsed = JSON.parse(json);
+
+    // get out test name
+    if (!Ember.empty(jsonParsed.name)){
+      testName =  jsonParsed.name.trim();
+    }
+
+    // get out results
+    if (jsonParsed.result && !Ember.empty(jsonParsed.result)){
+        jsonParsed.result.forEach(function(item){
```
Comment (mtwestra) [inline Dashboard/app/js/lib/views/views.js]: done

File: Dashboard/app/js/lib/views/views.js (hunk 26 above)
Comment (mtwestra) [inline Dashboard/app/js/lib/views/views.js]: done

File: GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java (hunk 28 above)
Comment (mtwestra) [inline GAE/src/com/gallatinsystems/survey/dao/CaddisflyResourceDao.java]: done

File: Dashboard/app/js/lib/views/surveys/question-view.js (hunk 34 above)
Comment (iperdomo) [inline Dashboard/app/js/lib/views/surveys/question-view.js]: Try: 
`FLOW.caddisflyResourceControl.get('content').findProperty('uuid', FLOW.selectedControl.selectedQuestion.get('caddisflyResourceUuid'))`

>    Returns the first item with a property matching the passed value.  You
>    can pass an optional second argument with the target value.  Otherwise
>    this will match any property that evaluates to true.
>    This method works much like the more generic find() method.

https://github.com/akvo/akvo-flow/blob/v1.9.10/Dashboard/app/js/vendor/ember-1.0.0.pre-2-36.js#L6954-L6967

Comment (mtwestra) [inline Dashboard/app/js/lib/views/surveys/question-view.js]: done
